src/main/java/htsjdk/variant/vcf/VCFIteratorBuilder.java

```diff
@@ -0,0 +1,204 @@
+* OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+* NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+* HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+* WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+* FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR
+* THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+*/
+package htsjdk.variant.vcf;
+
+import java.io.BufferedInputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Path;
+import java.util.zip.GZIPInputStream;
+
+import htsjdk.samtools.seekablestream.SeekablePathStream;
+import htsjdk.samtools.util.AbstractIterator;
+import htsjdk.samtools.util.CloserUtil;
+import htsjdk.samtools.util.IOUtil;
+import htsjdk.tribble.readers.AsciiLineReader;
+import htsjdk.tribble.readers.AsciiLineReaderIterator;
+import htsjdk.tribble.readers.PositionalBufferedStream;
+import htsjdk.tribble.util.ParsingUtils;
+import htsjdk.variant.bcf2.BCF2Codec;
+import htsjdk.variant.bcf2.BCFVersion;
+import htsjdk.variant.variantcontext.VariantContext;
+
+/**
+ * A Class building {@link VCFReaderIterator}
+ * 
+ * Example:
+ * 
+ * <pre>
+ * VCFIterator r = new VCFIteratorBuilder().open(System.in);
+ * while (r.hasNext()) {
+ *     System.out.println(r.next());
+ * }
+ * r.close();
+ * </pre>
+ * 
+ * @author Pierre Lindenbaum / @yokofakun
+ * @see {@link VCFIterator}
+ *
+ */
+
+public class VCFIteratorBuilder {
+
+    /**
+     * creates a VCF iterator from an input stream It detects if the stream is a
+     * BCF stream or a GZipped stream.
+     * 
+     * @param in inputstream
+     * @return the VCFIterator
+     * @throws IOException
+     */
+    public VCFIterator open(final InputStream in) throws IOException {
+        if (in == null) {
+            throw new IllegalArgumentException("input stream is null");
+            }
+        // buffer must be large enough to contain the BCF header (+ min/max version) and/or GZIP signature
+        final int bufferSize = Math.max( BCFVersion.MAGIC_HEADER_START.length + 2*Byte.BYTES , IOUtil.GZIP_SIGNATURE.length);
+        // try to read the BCF header
+        final BufferedInputStream bufferedinput = new BufferedInputStream(in,bufferSize);
+        bufferedinput.mark(bufferSize);
+        final BCFVersion bcfVersion = BCFVersion.readBCFVersion(bufferedinput);
+        bufferedinput.reset();
+
+        if (bcfVersion != null) {
+            //this is BCF
+            return new BCFInputStreamIterator(bufferedinput);
+        } else {
+            //
+            //this is VCF or VCF.gz, gunzip the input stream if needed
+            return new VCFReaderIterator( IOUtil.isGZIPInputStream(bufferedinput) ? 
+                    new GZIPInputStream(bufferedinput) : bufferedinput );
+        }
+    }
+
+    /**
+     * creates a VCF iterator from a URI It detects if the stream is a BCF
```

Comment:
Why? `ParsingUtils.openInputStream()` handles more than just files (it handles http URIs, as well as URIs with schemes supported by NIO plugins)